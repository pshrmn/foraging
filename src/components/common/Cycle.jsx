import React from 'react'

import { NeutralButton } from 'components/common/Buttons';

/*
 * Given an array of items, a Cycle uses two buttons to
 * allow you to cycle through the items in the array. The
 * Cycle only handles controlling the index that should
 * be shown, not displaying the actual items. A callback
 * is used to let the component in charge of displaying
 * items know the index of the one to show.
 *
 * When incrementing, automatically returns to 0 when end of
 * list has been reached. When decrementing, automatically
 * jumps to end of list after 0.
 */
const Cycle = ({ index, count, setIndex }) => {
  const nextIndex = (index+1) % count;
  const prevIndex = ((index-1) + count) % count;
  return (
    <div>
      <NeutralButton
        click={() => { setIndex(prevIndex); }}
        text='<' />
      {index + 1} / {count}
      <NeutralButton
        click={() => { setIndex(nextIndex); }}
        text='>' />
    </div>
  );
}

export default Cycle;
