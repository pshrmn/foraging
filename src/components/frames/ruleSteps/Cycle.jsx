import React from "react"

const Cycle = React.createClass({
  previousElement: function() {
    const { setIndex, index, count } = this.props;
    // add length then modulus because javascript mod negative stays negative
    const prev = ((index - 1) + count) % count;
    setIndex(prev);
  },
  nextElement: function() {
    const { setIndex, index, count } = this.props;
    const next = (index + 1) % count;
    setIndex(next);
  },
  render: function() {
    const { index, count } = this.props;
    return (
      <div>
        <button onClick={this.previousElement}>
          {"<"}
        </button>
        {index} / {count-1}
        <button onClick={this.nextElement}>
          {">"}
        </button>
      </div>
    );
  }
});

export default Cycle;
