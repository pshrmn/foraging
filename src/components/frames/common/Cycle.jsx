import React from 'react'

import { NeutralButton } from '../../common/Buttons';

export default function Cycle(props) {
  const { index, count, setIndex } = props;
  const nextIndex = (index+1) % count;
  const prevIndex = ((index-1) + count) % count;
  return (
    <div>
      <NeutralButton
        click={() => { setIndex(prevIndex); }}
        text='<' />
      {index + 1} / {count}
      <NeutralButton
        click={() => { setIndex(nextIndex); }}
        text='>' />
    </div>
  );
}
