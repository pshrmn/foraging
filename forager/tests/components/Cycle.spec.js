import React from 'react';
import renderer from 'react-test-renderer';
import Cycle from 'components/common/Cycle';

describe('Cycle', () => {
  it('create a Cycle component', () => {
    const component = renderer.create(
      <Cycle index={0} count={10}/>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('calls setIndex when left button is clicked', () => {
    const spy = jest.fn();
    const component = renderer.create(
      <Cycle index={5} count={10} setIndex={spy} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    // clicking the left button should call the function, passing
    // it index - 1
    tree.children[0].props.onClick()
    expect(spy.mock.calls.length).toBe(1)
    expect(spy.mock.calls[0]).toEqual([4])
  });

  it('calls setIndex when right button is clicked', () => {
    const spy = jest.fn();
    const component = renderer.create(
      <Cycle index={5} count={10} setIndex={spy} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    // clicking the left button should call the function, passing
    // it index + 1
    tree.children[4].props.onClick()
    expect(spy.mock.calls.length).toBe(1)
    expect(spy.mock.calls[0]).toEqual([6])
  });
});
