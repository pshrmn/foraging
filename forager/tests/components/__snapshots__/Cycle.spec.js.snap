// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cycle calls setIndex when left button is clicked 1`] = `
<div
  className="cycle"
>
  <button
    className=""
    disabled={false}
    onClick={[Function]}
    title=""
    type="button"
  >
    &lt;
  </button>
  6
   / 
  10
  <button
    className=""
    disabled={false}
    onClick={[Function]}
    title=""
    type="button"
  >
    &gt;
  </button>
</div>
`;

exports[`Cycle calls setIndex when right button is clicked 1`] = `
<div
  className="cycle"
>
  <button
    className=""
    disabled={false}
    onClick={[Function]}
    title=""
    type="button"
  >
    &lt;
  </button>
  6
   / 
  10
  <button
    className=""
    disabled={false}
    onClick={[Function]}
    title=""
    type="button"
  >
    &gt;
  </button>
</div>
`;

exports[`Cycle create a Cycle component 1`] = `
<div
  className="cycle"
>
  <button
    className=""
    disabled={false}
    onClick={[Function]}
    title=""
    type="button"
  >
    &lt;
  </button>
  1
   / 
  10
  <button
    className=""
    disabled={false}
    onClick={[Function]}
    title=""
    type="button"
  >
    &gt;
  </button>
</div>
`;
